src/pycdr/pycdr/types.py

```diff
@@ -62,95 +62,164 @@ def __repr__(self):
 
+    def __hash__(self) -> int:
+        return (241813571056069 * self.length) ^ hash(self.subtype)
 
-class ArrayGenerator:
-    def __getitem__(self, tup):
-        if type(tup) != tuple or len(tup) != 2 or type(tup[1]) != int:
-            return TypeError("An array takes two arguments: a type and a constant length.")
-        return Annotated[List[tup[0]], ArrayHolder(*tup)]
+    __str__ = __repr__
 
 
-class SequenceHolder:
-    def __init__(self, type, max_length):
-        self.type = type
-        self.max_length = max_length
+class sequence(metaclass=GetItemSupportMeta):
+    @classmethod
+    def __class_getitem__(cls, tup):
+        if type(tup) != tuple:
+            tup = (tup,)
+
+        if len(tup) not in [1, 2] or (len(tup) == 2 and type(tup[1]) != int):
+            raise TypeError("A sequence takes a subtype and an optional maximum length.")
+        if len(tup) > 1 and (tup[1] <= 0 or tup[1] > 65535):
+            return TypeError("Sequence max length should be between 0 and 65536.")
+        return Annotated[Sequence[tup[0]], cls(*tup)]
+
+    def __init__(self, subtype: Type, max_length: Optional[int] = None) -> None:
+        self.subtype: Type = subtype
+        self.max_length: Optional[int] = max_length
 
     def __repr__(self) -> str:
         if self.max_length:
-            return f"sequence[{self.type}, {self.max_length}]"
+            return f"sequence[{_type_repr(self.subtype)}, {self.max_length}]"
         else:
-            return f"sequence[{self.type}"
+            return f"sequence[{_type_repr(self.subtype)}]"
 
+    def __eq__(self, o: object) -> bool:
+        return isinstance(o, sequence) and o.subtype == self.subtype and o.max_length == self.max_length
 
-class SequenceGenerator:
-    def __getitem__(self, tup):
-        if type(tup) != tuple or len(tup) != 2 or type(tup[1]) != int:
-            if type(tup) == tuple:
-                return TypeError("A sequence takes one or two arguments: a type and an optional max length.")
-            return Annotated[List[tup], SequenceHolder(tup, None)]
-        if tup[1] <= 0 or tup[1] > 65535:
-            return TypeError("Sequence max length should be between 0 and 65536.")
-        return Annotated[List[tup[0]], SequenceHolder(*tup)]
+    def __hash__(self) -> int:
+        if self.max_length:
+            return (406820761152607 * self.max_length) ^ hash(self.subtype)
+        else:
+            return 406820761152609 ^ hash(self.subtype)
+
+    __str__ = __repr__
 
 
-class BoundStringHolder:
-    def __init__(self, max_length):
-        self.max_length = max_length
+class bounded_str(metaclass=GetItemSupportMeta):
+    @classmethod
+    def __class_getitem__(cls, tup):
+        if type(tup) != tuple:
+            tup = (tup,)
+
+        if len(tup) != 1 or type(tup[0]) != int:
+            raise TypeError("A bounded str takes one argument, a maximum length.")
+        if tup[0] <= 0 or tup[0] > 18446744073709551615:
```

Comment:
Where does the large constant come from? That's probably like a `SIZE_MAX` or something? Wouldn't it be way easier if we'd get that from the C layer? I know they're defines not symbols, but adding a `const` to the `clayer` that we can use here wouldn't be too much work right? Same for things like DDS_INFINITY etc... Would that be an option?

Reply:
I think I could actually remove this check, if you try to set a string limit of what in quick estimation are several exabytes then it will go wrong in other places.
I will move DDS_INFINITY into the `clayer`, I believe I can easily add it with `PyModule_AddIntMacro`.